### PR TITLE
Xiangyu/excution policy

### PR DESCRIPTION
**Pull request by Xiangyu-Hu**
1. The main interface function for computing in particle dynamics is now exec() only.
By introduce the concept of execution policy, running a method for particle dynamics in sequenced or parallel will be decide at the constructing process. 
This revision is for including vectorization codes for which we may need using different local dynamics functions such as interaction().
2. I have deleted the unused and untested functions for mirror boundary condition.

### CODE REVIEW

**Review by DongWuTUM — APPROVED** [review]

File: SPHINXsys/src/shared/particle_dynamics/execution_policy.h [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        {
+        public:
+            inline constexpr static ParallelUnsequencedPolicy generatePolicy()
+            {
+
+                return ParallelUnsequencedPolicy{};
+            };
+        };
+    }
+}
```
[the remark this answers is not in the document]
Comment (Xiangyu-Hu) [reply SPHINXsys/src/shared/particle_dynamics/execution_policy.h:74]: There are conflicts between the current standard library and the TBB used in library.

File: SPHINXsys/src/shared/particle_dynamics/fluid_dynamics/fluid_dynamics_complex.h [hunk 2]
```diff
@@ -179,7 +184,8 @@ namespace SPH
 			BaseIntegration1stHalfWithWall(Args &&...args)
 				: InteractionWithWall<BaseIntegration1stHalfType>(std::forward<Args>(args)...){};
 			virtual ~BaseIntegration1stHalfWithWall(){};
-			void interaction(size_t index_i, Real dt = 0.0);
+			
+			inline void interaction(size_t index_i, Real dt = 0.0);
```
[the remark this answers is not in the document]
Comment (Xiangyu-Hu) [reply SPHINXsys/src/shared/particle_dynamics/fluid_dynamics/fluid_dynamics_complex.h:188]: Yes. I can delete these keywords later.

File: SPHINXsys/src/shared/particle_dynamics/execution_policy.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (FabienPean-Virtonomy) [reply SPHINXsys/src/shared/particle_dynamics/execution_policy.h:74]: I meant same syntax structure as in the standard. `generatePolicy` is not needed.

```cpp
    namespace execution
    {
        class SequencedPolicy {};

        class UnsequencedPolicy {};

        class ParallelPolicy {};

        class ParallelUnsequencedPolicy {};

        inline constexpr auto seq = SequencedPolicy{};
        inline constexpr auto unseq = UnsequencedPolicy{};
        inline constexpr auto par = ParallelPolicy{};
        inline constexpr auto par_unseq = ParallelUnsequencedPolicy{};
    }
}
```

[the remark this answers is not in the document]
Comment (Xiangyu-Hu) [reply SPHINXsys/src/shared/particle_dynamics/execution_policy.h:74]: I did not found that these can be called in a class member function.

[the remark this answers is not in the document]
Comment (FabienPean-Virtonomy) [reply SPHINXsys/src/shared/particle_dynamics/execution_policy.h:74]: ?

**Review by FabienPean-Virtonomy — COMMENTED** [review]
xsimd can be installed with vcpkg instead as a submodule.

File: SPHINXsys/src/shared/particle_dynamics/particle_dynamics_algorithms.h [hunk 3]
```diff
@@ -147,24 +141,14 @@ namespace SPH
...
-		{
-			this->setupDynamics(dt);
-			ReturnType temp = particle_parallel_reduce(
-				this->identifier_.LoopRange(), this->Reference(), this->getOperation(),
-				[&](size_t i) -> ReturnType
-				{ return this->reduce(i, dt); });
+			ReturnType temp = particle_reduce(ExecutionPolicy::generatePolicy(),
+											  this->identifier_.LoopRange(), this->Reference(), this->getOperation(),
+											  [&](size_t i) -> ReturnType
+											  { return this->reduce(i, dt); });
```
Comment (Xiangyu-Hu) [inline SPHINXsys/src/shared/particle_dynamics/particle_dynamics_algorithms.h:151]: ExecutionPolicy::generatePolicy() needs to the same.

File: SPHINXsys/src/shared/particle_dynamics/execution_policy.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (Xiangyu-Hu) [reply SPHINXsys/src/shared/particle_dynamics/execution_policy.h:74]: I means the these function has different names and can not be used in the member function as it requires the same name belong a template class.
https://github.com/Xiangyu-Hu/SPHinXsys/pull/217/files#r1133906012

[the remark this answers is not in the document]
Comment (FabienPean-Virtonomy) [reply SPHINXsys/src/shared/particle_dynamics/execution_policy.h:74]: `ExecutionPolicy()`

[the remark this answers is not in the document]
Comment (Xiangyu-Hu) [reply SPHINXsys/src/shared/particle_dynamics/execution_policy.h:74]: Ok. Thanks.

**Comment by FabienPean-Virtonomy** [comment]
> I have tried and found, may be I did something wrong, not all functions can be used. Error message of not defined function shows up. So that I am using the submodule as it is header only anyway.

I made a commit to add the dependency, see https://github.com/Xiangyu-Hu/SPHinXsys/pull/217/commits/215fefde50a17ff438e9f1c5fd411688df127444

**Comment by Xiangyu-Hu** [comment]
> > I have tried and found, may be I did something wrong, not all functions can be used. Error message of not defined function shows up. So that I am using the submodule as it is header only anyway.
> 
> I made a commit to add the dependency, see [215fefd](https://github.com/Xiangyu-Hu/SPHinXsys/commit/215fefde50a17ff438e9f1c5fd411688df127444)

That is what I did exactly, and it did not work for me. Anyway, in this pull request, the xsimd code has not been included. 
You can simply neglect the submodule.

**Pull request merged**